packages/components/src/components/combobox/combobox.component.ts

```diff
@@ -0,0 +1,963 @@
+    this.setCurrentOption(null);
+  }
+
+  @watch('open', { waitUntilFirstUpdate: true })
+  async handleOpenChange() {
+    if (this.open && !this.disabled) {
+      if (this.filteredOptions.length === 0) {
+        // Don't open the listbox if there are no options
+        this.open = false;
+        this.emit('syn-error');
+        return;
+      }
+
+      // Show
+      this.emit('syn-show');
+      this.addOpenListeners();
+
+      await stopAnimations(this);
+      this.listbox.hidden = false;
+      this.popup.active = true;
+
+      const { keyframes, options } = getAnimation(this, 'combobox.show', { dir: this.localize.dir() });
+      await animateTo(this.popup.popup, keyframes, options);
+
+      this.emit('syn-after-show');
+    } else {
+      this.setCurrentOption(null);
+      this.displayInput.removeAttribute('aria-activedescendant');
+      // Hide
+      this.emit('syn-hide');
+      this.removeOpenListeners();
+
+      await stopAnimations(this);
+      const { keyframes, options } = getAnimation(this, 'combobox.hide', { dir: this.localize.dir() });
+      await animateTo(this.popup.popup, keyframes, options);
+      this.listbox.hidden = true;
+      this.popup.active = false;
+
+      this.emit('syn-after-hide');
+    }
+  }
+
+  /**
+   * Shows the listbox. If it is not possible to open the listbox, because there are no
+   * appropriate filtered options, a syn-error is emitted and the listbox stays closed.
+   */
+  async show() {
+    if (this.open || this.disabled) {
+      this.open = false;
+      return undefined;
+    }
+
+    this.open = true;
+    return Promise.race([waitForEvent(this, 'syn-after-show'), waitForEvent(this, 'syn-error')]);
+  }
+
+  /** Hides the listbox. */
+  async hide() {
+    if (!this.open || this.disabled) {
+      this.open = false;
+      return undefined;
+    }
+
+    this.open = false;
+    return waitForEvent(this, 'syn-after-hide');
+  }
+
+  /**
+   * Checks for validity but does not show a validation message.
+   * Returns `true` when valid and `false` when invalid.
+   */
+  checkValidity() {
+    return this.valueInput.checkValidity();
+  }
+
+  /** Gets the associated form, if one exists. */
+  getForm(): HTMLFormElement | null {
+    return this.formControlController.getForm();
+  }
+
+  /** Checks for validity and shows the browser's validation message if the control is invalid. */
+  reportValidity() {
+    return this.valueInput.reportValidity();
+  }
+
+  /** Sets a custom validation message. Pass an empty string to restore validity. */
+  setCustomValidity(message: string) {
+    this.valueInput.setCustomValidity(message);
+    this.formControlController.updateValidity();
+  }
+
+  /** Sets focus on the control. */
+  focus(options?: FocusOptions) {
+    this.displayInput.focus(options);
+  }
+
+  /** Removes focus from the control. */
+  blur() {
+    this.displayInput.blur();
+  }
+
+  private createComboboxOptionsFromQuery(queryString: string) {
```

Comment:
The code is pretty hard to read in my opinion, maybe we can come up with an easier version, something like this (not tested):

```typescript
private createComboboxOptionsFromQuery(queryString: string) {
    this.filteredOptions = this.getSlottedOptions()
      .filter(o => this.filter(o, queryString) || queryString === '')
      .map(o => {
        const clonedOption = option.cloneNode(true) as SynOption;
        // Check if the option has a syn-optgroup as parent
        const hasOptgroup = option.parentElement?.tagName.toLowerCase() === 'syn-optgroup';

        if (!hasOptgroup) {
          return clonedOption;
        }

        const optgroup = option.parentElement as SynOptGroup;
        const filteredOptgroup = this.filteredOptions.find((el) => el.id === optgroup.id);
        // Check if the optgroup was already added to the filteredOptions
        if (!filteredOptgroup) {
          const clonedOptgroup = option.parentElement.cloneNode() as SynOptGroup;
          clonedOptgroup.appendChild(clonedOption);
          return clonedOptgroup;
        } else {
          filteredOptgroup?.appendChild(clonedOption);
        }
      });
  }
```

Reply:
I needed to adapt some things, but I tried to make it more readable :)